### PR TITLE
[guiinfo] Major refactoring of CGUIInfoManager

### PR DESCRIPTION
**Pull request by ksooo**
I ran mad and invested 3 weeks into breaking the GUIInfoManager monster into imo managable pieces.

Main idea is to introduce a `guiinfo data provider` per Kodi component to decouple the manager from the data sources. See commit 1 to get the idea.

Now we have a provider for video, one for music, for pictures, pvr, addons, ... 

Also, this is a first step into direction to change guiinfo manager from 'pull' to 'push' paradigm. Now, the components can be changed one after the other.

This PR should not contain any functional changes. I did my best to ensure everything is still working as before.

I runtime tested this very careful, but I'm pretty sure I broke some edge cases. ;-)

### CODE REVIEW

**Comment by FernetMenta** [comment]
could you please squash all providers that pull info from players into a single one?

ingo related to playlists should not be in player because players does not know anthing about playlists

**Comment by ksooo** [comment]
> could you please squash all providers that pull info from players into a single one?

 not sure what exactly you want me to do? what does "pull info from players" mean code-wise?

> info related to playlists should not be in player because players does not know anthing about playlists

 okay, but where should playlists go?

**Comment by ksooo** [comment]
>>  could you please squash all providers that pull info from players into a single one?

> not sure what exactly you want me to do? what does "pull info from players" mean code-wise?

Is it calls to `CApplication::GetAppPlayer`?

**Comment by FernetMenta** [comment]
> not sure what exactly you want me to do? what does "pull info from players" mean code-wise?

everything that starts with PLAYER_ or VIDEOPLAYER_

> okay, but where should playlists go?

maybe some PlaylistPlayer info class

**Comment by ksooo** [comment]
> everything that starts with PLAYER_ or VIDEOPLAYER_

 what about MUSICPLAYER_*?

**Comment by ksooo** [comment]
> maybe some PlaylistPlayer info class

 it has to go into a guiinfo provider. which one? a new - playlist - one?

**Comment by FernetMenta** [comment]
> what about MUSICPLAYER_*?

most info comes from PlaylistPlayer

> it has to go into a guiinfo provider. which one? a new - playlist - one?

maybe one provide for player named player and the rest goes to PlayerEx or something

**Comment by DaveTBlake** [comment]
FWIW splitting into provider by where the info comes from e.g. VideoInfoTag, MusicInfoTag etc. like you have @ksooo seems a smart idea to me. It is so much clearer, and more maintainable to have the processing for both MUSICPLAYER_ and the similar LISTITEM_ together.

File: xbmc/guiinfo/GUIInfoLabels.h [hunk 1]
```diff
@@ -799,30 +860,21 @@
...
-#define MUSICPLAYER_PROPERTY_OFFSET       1300  // 200 id's reserved for musicplayer props.
-#define LISTITEM_ART_OFFSET               1500  // 200 id's reserved for listitem art.
-#define LISTITEM_RATING_OFFSET      1700 // 200 id's reserved for listitem ratings.
-#define LISTITEM_VOTES_OFFSET       1900 // 200 id's reserved for listitem votes.
-#define LISTITEM_RATING_AND_VOTES_OFFSET  2100 // 200 id's reserved for listitem ratingandvotes.
-#define LISTITEM_DURATION_OFFSET    2300 // 200 id's reserved for listitem duration.
+#define LISTITEM_END                (LISTITEM_START + 2500)
 
-#define CONDITIONAL_LABEL_START       LISTITEM_END + 1 // 37701
+#define CONDITIONAL_LABEL_START       (LISTITEM_END + 1) // 37501
```
Comment (ronie) [inline xbmc/guiinfo/GUIInfoLabels.h]: while you're at it, mind increasing CONDITIONAL_LABEL_END to 39999 ?

ref: https://forum.kodi.tv/showthread.php?tid=330522

Comment (ksooo) [reply xbmc/guiinfo/GUIInfoLabels.h]: np, will do

**Comment by ksooo** [comment]
>  It is so much clearer, and more maintainable to have the processing for both MUSICPLAYER_ and the similar LISTITEM_ together.

 yep, basic idea is not grouping by label prefix, but by "data source".

File: xbmc/guiinfo/GUIInfoLabels.h (hunk 1 above)
Comment (ksooo) [reply xbmc/guiinfo/GUIInfoLabels.h]: done, new commit for this.

**Comment by FernetMenta** [comment]
@ksooo when we move from pull to push, the sources will push to a single cache. Means we need a 1:1 relationship between source and provider. 
There could be secondary providers without a single source like MUSICPLAYER_. They can pull their infos from other providers and should not have a direct source like player.

**Comment by ksooo** [comment]
I want to make a suggestion: Lets decide for this PR that separating GUIInfoManager functionality into different guiinfo providers is the way we want to go (or not). If we are fine with this general approach, let's test that the implementation provided with the PR works in general and does not introduce regressions.

Discussing which label belongs into which provider and which providers we need can take a long time and adjustments to this can be done any time later without bad side effects. So, let's move this fine tuning into followup PRs. I volonteer to take care of the resulting followup tasks. Otherwise I see 'chances' that this PR will never go in (and that discussion will become frustrating very quickly). And believe me, I don't want to rebase this monster forever.

So, give me your +1 / -1 on my suggestion and I will obey the outcome. ;-)

**Review by Paxxi — APPROVED** [review]
I don't really have the time or energy to do a through review but I think the high level stuff looks good. Very much needed work

Some nitpicks that you can consider suggestions only :) I would have put it into KODI::GUIINFO and I would've made the helper functions free, no real need to wrap them in a class, they're already enclosed in the namespace.

**Comment by FernetMenta** [comment]
I am fine with moving labels around later. But since this PR creates a lot of new files, folder guiinfo should be moved to guilib.

File: xbmc/guiinfo/CMakeLists.txt [hunk 2]
```diff
@@ -1,5 +1,9 @@
+set(SOURCES GUIInfoProviders.cpp)
```
[the remark this answers is not in the document]
Comment (FernetMenta) [reply xbmc/guiinfo/CMakeLists.txt]: currently guiinfo has only 2 files and this creates a lot more. let's avoid future work by putting new files in correct folder.

**Comment by ksooo** [comment]
@Paxxi thanke for taking a look.

> I would have put it into KODI::GUIINFO and I would've made the helper functions free, no real need to wrap them in a class, they're already enclosed in the namespace.

 Good points. I will change that.

**Comment by FernetMenta** [comment]
> Do you want the whole guiinfo folder to be a subfolder of guilb? I'm fine with that, just asking.

yes, I think this makes sense. could be done by an addition commit. no need to touch every single one.

**Comment by ksooo** [comment]
> could be done by an addition commit. no need to touch every single one.

 the latter would indeed be a rebase nightmare.

**Comment by ksooo** [comment]
@FernetMenta files moved from "guiinfo/" to "guilib/guiinfo/"

@Paxxi namespace now `KODI::GUILIB::GUIINFO`; `CGUIInfoHelper` removed in favour to free functions.

**Comment by ronie** [comment]
bug reports: https://forum.kodi.tv/showthread.php?tid=330696

**Comment by ksooo** [comment]
@ronie @MilhouseVH @Razzeee  latest commit fixes the non-working covers in the info dialogs. So, if you want to kick another test build... ;-)

**Comment by Razzeee** [comment]
http://jenkins.kodi.tv/view/Helpers/job/BuildMulti-PR-Manually/189/

**Comment by ksooo** [comment]
Many thanks to all testers. No more open issues. If nobody objects I will merge this PR the next days.

**Comment by ksooo** [comment]
jenkins build this please

**Comment by ksooo** [comment]
jenkins build this please

**Comment by ksooo** [comment]
No more open issues. Many thanks to all testers for supporting me.

**Pull request merged**